full.rollbar/src/full/rollbar/core.clj

```diff
@@ -0,0 +1,79 @@
+(ns full.rollbar.core
+  (:require [full.http.client :refer [req>]]
+            [full.async :refer [go-try <?? <?]]
+            [full.core.sugar :refer [when->]]
+            [camelsnake.core :refer [->snake_case]]
+            [full.core.log :as log])
+  (:import (java.net InetAddress)))
+
+(defn- rollbar-req>
+  [body]
+  (req> {:base-url "https://api.rollbar.com"
+         :resource "api/1/item/"
+         :method :post
+         :body body
+         :body-json-key-fn ->snake_case}))
+
+(defn root-exception
+  [ex]
+  (loop [ex ex]
+    (if-let [cause (.getCause ex)]
+      (recur cause)
+      ex)))
+
+(defn error-msg
+  [exc]
+  (let [message (.getMessage exc)]
+    (if (empty? message)
+      (.toString (.getClass exc))
+      message)))
+
+(defn- frame
+  [row]
+  {:filename (-> row .getFileName)
+   :lineno (-> row .getLineNumber)
+   :method (str (-> row .getClassName) "/" (-> row .getMethodName))})
+
+(defn- frames
+  [ex]
+  (->> ex (.getStackTrace) (map frame)))
+
+(defn- exception
+  [ex]
+  {:class (-> ex .getClass .toString)
+   :message (error-msg ex)})
+
+(defn trace-payload
+  [ex]
+  (let [root-ex (root-exception ex)]
+    {:frames (frames root-ex)
+     :exception (exception root-ex)}))
+
+(defn request-payload
+  [req]
+  {:method (clojure.string/upper-case (name (:request-method req)))
+   :url (:uri req)
+   :headers (:headers req)
+   :user_ip (:remote-addr req)})
+
+(defn rollbar-payload
+  [host access-token environment]
+  {:access_token access-token
+   :data {:environment environment
+          :host host
+          :language "clojure"
+          :body {}}})
+
+(def host (.getHostAddress (InetAddress/getLocalHost)))
+
+(defn report>
+  [access-token environment ex & [request]]
+  (go-try
+    (if (and access-token environment)
```

Comment:
Also, if values aren't provided in pre, won't it throw an exception?  I'd hate to fail requests for an improperly configured rollbar client.  I'd prefer to warn gratuitously.